util/resolver/resolver.go

```diff
@@ -1,54 +1,116 @@
 package resolver
 
 import (
-	"math/rand"
-	"strings"
+	"context"
+	"crypto/tls"
+	"net/http"
+	"time"
 
+	"github.com/containerd/containerd/remotes"
 	"github.com/containerd/containerd/remotes/docker"
-	"github.com/docker/distribution/reference"
+	"github.com/moby/buildkit/session"
+	"github.com/moby/buildkit/session/auth"
 	"github.com/moby/buildkit/util/tracing"
 )
 
 type RegistryConf struct {
 	Mirrors   []string
 	PlainHTTP *bool
+	Insecure  *bool
 }
 
-type ResolveOptionsFunc func(string) docker.ResolverOptions
-
-func NewResolveOptionsFunc(m map[string]RegistryConf) ResolveOptionsFunc {
-	return func(ref string) docker.ResolverOptions {
-		def := docker.ResolverOptions{
-			Client: tracing.DefaultClient,
+func fillInsecureOpts(host string, c RegistryConf, h *docker.RegistryHost) {
+	if c.PlainHTTP != nil && *c.PlainHTTP {
+		h.Scheme = "http"
+	} else if c.Insecure != nil && *c.Insecure {
+		h.Client = &http.Client{
+			Transport: tracing.NewTransport(&http.Transport{TLSClientConfig: &tls.Config{InsecureSkipVerify: true}}),
```

Comment:
thanks god we have this patch. we are using version 0.6.2 and use http2 to push/pull image if there is no TLSClientConfig. The http2 flow control is killing me which makes the push very slow.

anyway, nice to have this patch!

Reply:
@fuweid I think this requires more optimization actually. Afaics the issue with the array model is that it doesn't lock to the specific config when other fail. So it keeps retrying the ones that will fail for every request making things like mirrors inefficient.